src/libply/arch/mips.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Ye Jiaqiang <yejq.jiaqiang@gmail.com>
+ *
+ * SPDX-License-Identifier: GPL-2.0
+ */
+
+#include <assert.h>
+
+#include <ply/internal.h>
+
+#define arch_typedef(_a, _t) {					\
+		.ttype = T_TYPEDEF,				\
+		.tdef = { .name = #_a, .type = _t },		\
+	}
+
+struct type t_s8  = arch_typedef(s8,  &t_schar);
+struct type t_u8  = arch_typedef(u8,  &t_uchar);
+struct type t_s16 = arch_typedef(s16, &t_sshort);
+struct type t_u16 = arch_typedef(u16, &t_ushort);
+struct type t_s32 = arch_typedef(s32, &t_sint);
+struct type t_u32 = arch_typedef(u32, &t_uint);
+struct type t_s64 = arch_typedef(s64, &t_sllong);
+struct type t_u64 = arch_typedef(u64, &t_ullong);
+
+static int reg_fprint(struct type *t, FILE *fp, const void *data)
+{
+	return fprintf(fp, "%#lx", *((unsigned long *)data));
+}
+
+struct type t_reg_t = {
+	.ttype = T_TYPEDEF,
+	.tdef = {
+		.name = "reg_t",
+		.type = &t_ulong,
+	},
+
+	.fprint = reg_fprint,
+};
+
+#if defined(USE_MIPS_64BIT) && !defined(USE_MIPS_N32)
+  #define USE_MIPS_N32 1
+#endif
+
+struct tfield f_pt_regs_fields[] = {
+#ifndef USE_MIPS_64BIT
+	/* pad for 32-bit system; refer to linux/arch/mips/include/asm/ptrace.h */
+	{ .name = "uarg0",        .type = &t_reg_t },
+	{ .name = "uarg1",        .type = &t_reg_t },
+	{ .name = "uarg2",        .type = &t_reg_t },
+	{ .name = "uarg3",        .type = &t_reg_t },
+	{ .name = "uarg4",        .type = &t_reg_t },
+	{ .name = "uarg5",        .type = &t_reg_t },
+	{ .name = "uarg6",        .type = &t_reg_t },
+	{ .name = "uarg7",        .type = &t_reg_t },
+#endif
```

Comment:
Since there is no uprobe support yet this could be fine for now, but just for my understanding: Is the layout for `struct pt_regs` different for a kernel thread and a user thread? I.e. when we add uprobe support do we need to have another type for that?

Reply:
There is other `struct pt_regs` defined in `arch/mips/include/uapi/asm/ptrace.h`, which does not have 8 padding userspace registers. The kernel documentation says it is specifically used for `ptrace`:
> This struct defines the registers as used by PTRACE_{GET,SET}REGS.
So I recon that the layout for kernel threads and user threads are identical on MIPS32. However, for MIPS64, 8 padding userspace registers are not present in `struct pt_regs` (accordding to `arch/mips/include/asm/ptrace.h`). If support for MIPS64 is to be dropped, the code can be simplified a lot.